src/platform/backends/qemu/dnsmasq_server.cpp

```diff
@@ -105,3 +135,21 @@ void mp::DNSMasqServer::release_mac(const std::string& hw_addr)
 
     dhcp_release.waitForFinished();
 }
+
+void mp::DNSMasqServer::check_dnsmasq_running()
+{
+    auto dnsmasq_pid = get_dnsmasq_pid(pid_file_path);
+    if (kill(dnsmasq_pid, 0) == 0)
+        return;
+
+    // exit_code == 2 signifies dnsmasq network-related error. See `man dnsmasq`.
+    throw std::runtime_error(
+        fmt::format("Multipass dnsmasq is not running.{}",
+                    (dnsmasq_exit_state.exit_code == 2) ? " Ensure nothing is using port 53." : ""));
```

Comment:
Unlikely, but if dnsmasq crashed, dnsmasq_exit_state.exit_code will not be set, so it will throw too

Reply:
Isn't that what we want?  I'm not I follow what you mean here.  We want to throw if `dnsmasq` is not running for whatever reason.

Reply:
Ok. This is what I dislike about exceptions, it's very hard to know what error handling was planned when reading the code.

Reply:
I redid this a bit, but it still throws here, so I think your dislike still applies.